File: cli/gh-teacher/internal/configrepo/teamdescription_test.go
package configrepo

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/foundation50/classroom50-cli-shared/contract"
)

// TestMarshalTeamDescription pins the classroom50/team/v1 encoding: schema
// always present, empty fields omitted, active omitted when true, secret only
// when valid.
func TestMarshalTeamDescription(t *testing.T) {
	t.Run("unlisted active classroom includes secret, omits active", func(t *testing.T) {
		got, err := MarshalTeamDescription("Intro CS", "Fall 2026", "a1b2c3d4", "", true)
		if err != nil {
			t.Fatalf("MarshalTeamDescription: %v", err)
		}
		var decoded map[string]any
		if err := json.Unmarshal([]byte(got), &decoded); err != nil {
			t.Fatalf("result is not valid JSON: %v (%q)", err, got)
		}
		if decoded["schema"] != contract.TeamSchemaV1 {
			t.Errorf("schema = %v, want %q", decoded["schema"], contract.TeamSchemaV1)
		}
		if decoded["name"] != "Intro CS" {
			t.Errorf("name = %v, want %q", decoded["name"], "Intro CS")
		}
		if decoded["term"] != "Fall 2026" {
			t.Errorf("term = %v, want %q", decoded["term"], "Fall 2026")
		}
		if decoded["secret"] != "a1b2c3d4" {
			t.Errorf("secret = %v, want %q", decoded["secret"], "a1b2c3d4")
		}
		if _, present := decoded["active"]; present {
			t.Errorf("active must be omitted when true, got %v", decoded["active"])
		}
	})

	t.Run("listed classroom omits secret", func(t *testing.T) {
		got, err := MarshalTeamDescription("Intro CS", "", "", "", true)
		if err != nil {
			t.Fatalf("MarshalTeamDescription: %v", err)
		}
		var decoded map[string]any
		if err := json.Unmarshal([]byte(got), &decoded); err != nil {
			t.Fatalf("invalid JSON: %v", err)
		}
		if _, present := decoded["secret"]; present {
			t.Errorf("secret must be omitted for a listed classroom, got %v", decoded["secret"])
		}
		if _, present := decoded["term"]; present {
			t.Errorf("empty term must be omitted, got %v", decoded["term"])
		}
	})

	t.Run("archived classroom writes active=false", func(t *testing.T) {
		got, err := MarshalTeamDescription("Intro CS", "", "", "", false)
		if err != nil {
			t.Fatalf("MarshalTeamDescription: %v", err)
		}
		var decoded map[string]any
		if err := json.Unmarshal([]byte(got), &decoded); err != nil {
			t.Fatalf("invalid JSON: %v", err)
		}
		active, ok := decoded["active"].(bool)
		if !ok || active {
			t.Errorf("active = %v, want false present", decoded["active"])
		}
	})

	t.Run("malformed secret is rejected, not persisted", func(t *testing.T) {
		if _, err := MarshalTeamDescription("Intro CS", "", "BAD-secret!", "", true); err == nil {
			t.Error("expected an error for a malformed secret, got nil")
		}
	})

	t.Run("valid pages_base_url is included", func(t *testing.T) {
		got, err := MarshalTeamDescription("Intro CS", "", "", "https://cs.example.edu/classroom50", true)
		if err != nil {
			t.Fatalf("MarshalTeamDescription: %v", err)
		}
		var decoded map[string]any
		if err := json.Unmarshal([]byte(got), &decoded); err != nil {
			t.Fatalf("invalid JSON: %v", err)
		}
		if decoded["pages_base_url"] != "https://cs.example.edu/classroom50" {
			t.Errorf("pages_base_url = %v, want %q", decoded["pages_base_url"], "https://cs.example.edu/classroom50")
		}
	})

	t.Run("malformed pages_base_url is dropped, not persisted", func(t *testing.T) {
		for _, bad := range []string{
			"http://cs.example.edu",
			"https://cs.example.edu/",
			"https://cs.example.edu/x?y=1",
			"cs.example.edu",
		} {
			got, err := MarshalTeamDescription("Intro CS", "", "", bad, true)
			if err != nil {
				t.Fatalf("MarshalTeamDescription(%q): %v", bad, err)
			}
			var decoded map[string]any
			if err := json.Unmarshal([]byte(got), &decoded); err != nil {
				t.Fatalf("invalid JSON: %v", err)
			}
			if _, present := decoded["pages_base_url"]; present {
				t.Errorf("pages_base_url %q must be dropped, got %v", bad, decoded["pages_base_url"])
			}
		}
	})

	t.Run("stays under the size budget", func(t *testing.T) {
		// A realistic worst case: long name + term + secret + custom base URL.
		// Must stay well under ~250 chars so the description never risks
		// GitHub's field limit.
		got, err := MarshalTeamDescription(
			"Introduction to Computer Science and Programming",
			"Fall Semester 2026",
			"a1b2c3d4e5f6g7h8",
			"https://pages.cs.example-university.edu/classroom50",
			true,
		)
		if err != nil {
			t.Fatalf("MarshalTeamDescription: %v", err)
		}
		if len(got) > 250 {
			t.Errorf("encoded description is %d chars, want <= 250: %q", len(got), got)
		}
	})
}

// sharedTeamDescriptionCasesPath locates the cross-language golden fixture, also
// consumed by the TS mirror (web/src/util/teamDescription.test.ts).
const sharedTeamDescriptionCasesPath = "../../../shared/testdata/team_description_cases.json"

// TestMarshalTeamDescription_SharedFixtureParity pins the Go encoding to the
// shared golden bytes so the Go writer and the TS marshalTeamDescription can't
// drift: a one-sided edit (e.g., HTML-escaping) fails on the other language's
// copy of these same cases. Byte-identity matters because the web reconcile
// compares the current description to the desired string for exact equality —
// any divergence makes the CLI and web perpetually overwrite each other.
func TestMarshalTeamDescription_SharedFixtureParity(t *testing.T) {
	raw, err := os.ReadFile(filepath.Clean(sharedTeamDescriptionCasesPath))
	if err != nil {
		t.Fatalf("read shared fixture: %v", err)
	}
	var doc struct {
		Cases []struct {
			Input struct {
				Name         string `json:"name"`
				Term         string `json:"term"`
				Secret       string `json:"secret"`
				PagesBaseURL string `json:"pages_base_url"`
				Active       bool   `json:"active"`
			} `json:"input"`
			Encoded string `json:"encoded"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("parse shared fixture: %v", err)
	}
	if len(doc.Cases) == 0 {
		t.Fatal("shared fixture has no cases")
	}
	for _, c := range doc.Cases {
		got, err := MarshalTeamDescription(c.Input.Name, c.Input.Term, c.Input.Secret, c.Input.PagesBaseURL, c.Input.Active)
		if err != nil {
			t.Fatalf("MarshalTeamDescription(%+v): %v", c.Input, err)
		}
		if got != c.Encoded {
			t.Errorf("MarshalTeamDescription(%+v) = %q, want %q (cross-language drift — update every copy in lockstep)", c.Input, got, c.Encoded)
		}
	}
}

File: web/src/util/teamDescription.test.ts
import { describe, expect, it } from "vitest"
import { readFileSync } from "node:fs"
import { fileURLToPath } from "node:url"
import { marshalTeamDescription, parseTeamDescription } from "./teamDescription"

describe("parseTeamDescription", () => {
  it("parses a valid v1 record", () => {
    const desc = JSON.stringify({
      schema: "classroom50/team/v1",
      name: "Intro CS",
      term: "Fall 2026",
      secret: "a1b2c3d4",
    })
    expect(parseTeamDescription(desc)).toEqual({
      schema: "classroom50/team/v1",
      name: "Intro CS",
      term: "Fall 2026",
      secret: "a1b2c3d4",
    })
  })

  it("returns no secret for a listed classroom", () => {
    const desc = JSON.stringify({ schema: "classroom50/team/v1", name: "CS" })
    const parsed = parseTeamDescription(desc)
    expect(parsed.name).toBe("CS")
    expect(parsed.secret).toBeUndefined()
  })

  it("drops a malformed secret rather than failing the parse", () => {
    const desc = JSON.stringify({
      schema: "classroom50/team/v1",
      name: "CS",
      secret: "BAD secret!",
    })
    const parsed = parseTeamDescription(desc)
    expect(parsed.name).toBe("CS")
    expect(parsed.secret).toBeUndefined()
  })

  it("parses a valid pages_base_url and drops a malformed one", () => {
    const valid = parseTeamDescription(
      JSON.stringify({
        schema: "classroom50/team/v1",
        name: "CS",
        pages_base_url: "https://cs.example.edu/classroom50",
      }),
    )
    expect(valid.pages_base_url).toBe("https://cs.example.edu/classroom50")

    const invalid = parseTeamDescription(
      JSON.stringify({
        schema: "classroom50/team/v1",
        name: "CS",
        pages_base_url: "http://cs.example.edu/",
      }),
    )
    expect(invalid.name).toBe("CS")
    expect(invalid.pages_base_url).toBeUndefined()
  })

  it("ignores unknown future fields (additive evolution)", () => {
    const desc = JSON.stringify({
      schema: "classroom50/team/v1",
      name: "CS",
      futureField: "x",
    })
    const parsed = parseTeamDescription(desc)
    expect(parsed.name).toBe("CS")
    expect(parsed.schema).toBe("classroom50/team/v1")
  })

  it("returns {} for a plain-text (pre-schema) description", () => {
    expect(parseTeamDescription("Students of CS101")).toEqual({})
  })

  it("returns {} for a wrong/absent schema sentinel", () => {
    expect(
      parseTeamDescription(JSON.stringify({ schema: "other", secret: "abcd" })),
    ).toEqual({})
  })

  it("returns {} for null, undefined, or empty", () => {
    expect(parseTeamDescription(null)).toEqual({})
    expect(parseTeamDescription(undefined)).toEqual({})
    expect(parseTeamDescription("")).toEqual({})
  })
})

describe("marshalTeamDescription", () => {
  it("encodes an unlisted active classroom with secret, omitting active", () => {
    const out = marshalTeamDescription({
      name: "Intro CS",
      term: "Fall 2026",
      secret: "a1b2c3d4",
      active: true,
    })
    expect(JSON.parse(out)).toEqual({
      schema: "classroom50/team/v1",
      name: "Intro CS",
      term: "Fall 2026",
      secret: "a1b2c3d4",
    })
  })

  it("omits empty name/term and the secret for a listed classroom", () => {
    const out = marshalTeamDescription({ name: "CS", active: true })
    expect(JSON.parse(out)).toEqual({
      schema: "classroom50/team/v1",
      name: "CS",
    })
  })

  it("drops a malformed secret rather than persisting it", () => {
    const out = marshalTeamDescription({
      name: "CS",
      secret: "BAD secret!",
      active: true,
    })
    expect(JSON.parse(out).secret).toBeUndefined()
  })

  it("emits active:false for an archived classroom", () => {
    const out = marshalTeamDescription({ name: "CS", active: false })
    expect(JSON.parse(out).active).toBe(false)
  })

  it("includes a valid pages_base_url", () => {
    const out = marshalTeamDescription({
      name: "CS",
      pagesBaseUrl: "https://cs.example.edu/classroom50",
      active: true,
    })
    expect(JSON.parse(out).pages_base_url).toBe(
      "https://cs.example.edu/classroom50",
    )
  })

  it("drops a malformed pages_base_url rather than persisting it", () => {
    for (const bad of [
      "http://cs.example.edu",
      "https://cs.example.edu/",
      "https://cs.example.edu/x?y=1",
      "cs.example.edu",
    ]) {
      const out = marshalTeamDescription({
        name: "CS",
        pagesBaseUrl: bad,
        active: true,
      })
      expect(JSON.parse(out).pages_base_url).toBeUndefined()
    }
  })

  it("round-trips through parseTeamDescription", () => {
    const out = marshalTeamDescription({
      name: "Intro CS",
      term: "Fall 2026",
      secret: "a1b2c3d4",
      active: true,
    })
    expect(parseTeamDescription(out)).toEqual({
      schema: "classroom50/team/v1",
      name: "Intro CS",
      term: "Fall 2026",
      secret: "a1b2c3d4",
    })
  })

  it("escapes <, >, &, U+2028, U+2029 to match Go's json.Marshal (byte-identical contract)", () => {
    // Go's json.Marshal escapes all five by default; JSON.stringify escapes
    // none. Without matching, the CLI and web would perpetually rewrite each
    // other's description for a classroom whose name/term contains them.
    const out = marshalTeamDescription({ name: "C++ & <Data>", active: true })
    expect(out).toContain("\\u0026")
    expect(out).toContain("\\u003c")
    expect(out).toContain("\\u003e")
    expect(out).not.toMatch(/[<>&]/)
    // Still valid JSON that parses back to the original name.
    expect(parseTeamDescription(out).name).toBe("C++ & <Data>")

    // Line/paragraph separators (a common paste vector) are the only other
    // divergence Go escapes; pin them explicitly.
    const seps = marshalTeamDescription({
      name: "a\u2028b\u2029c",
      active: true,
    })
    expect(seps).toContain("\\u2028")
    expect(seps).toContain("\\u2029")
    expect(seps).not.toMatch(/[\u2028\u2029]/)
    expect(parseTeamDescription(seps).name).toBe("a\u2028b\u2029c")
  })
})

// Cross-language byte-identity: the same golden cases the Go
// MarshalTeamDescription asserts (cli/gh-teacher/internal/configrepo/
// teamdescription_test.go). A drift on either side fails here — the web
// reconcile compares exact strings, so divergence causes perpetual rewrites.
describe("marshalTeamDescription — shared fixture parity", () => {
  const fixtureUrl = new URL(
    "../../../cli/shared/testdata/team_description_cases.json",
    import.meta.url,
  )
  const doc = JSON.parse(readFileSync(fileURLToPath(fixtureUrl), "utf8")) as {
    cases: {
      input: {
        name?: string
        term?: string
        secret?: string
        pages_base_url?: string
        active: boolean
      }
      encoded: string
    }[]
  }

  it("has cases", () => {
    expect(doc.cases.length).toBeGreaterThan(0)
  })

  for (const [i, c] of doc.cases.entries()) {
    it(`case ${i} encodes byte-identically to the Go writer`, () => {
      expect(
        marshalTeamDescription({
          name: c.input.name,
          term: c.input.term,
          secret: c.input.secret,
          pagesBaseUrl: c.input.pages_base_url,
          active: c.input.active,
        }),
      ).toBe(c.encoded)
    })
  }
})
